### PR TITLE
Reject commands if the preconditions are not met

### LINKED ISSUE #6204: Reject commands if the preconditions are not met
**Description**

Currently, we don't process events if the preconditions are not met (see `io.zeebe.engine.processing.bpmn.WorkflowInstanceStateTransitionGuard`). For example, when the workflow execution is interrupted by a cancel command or an interrupting flow. When applying event-sourcing to the stream processors, we need to reject the commands and can not ignore them. 

* reject commands if the preconditions are not met 

Reference:
* [ZEP: Concurrent Commands and Rejections](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md#concurrent-commands-and-rejections)

### PR DESCRIPTION
**Pull request by saig0**
## Description

* if the preconditions of a command are not met then write a rejection record on the stream 
  * it makes the fact more transparent for a consumer of the stream that the command is not processed and why
  * the rejection replaces a `debug` log statement that was written in these cases

## Related issues

closes #6204

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/test/java/io/zeebe/engine/processing/processinstance/ProcessInstanceCommandRejectionTest.java [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+
+public final class ProcessInstanceCommandRejectionTest {
+
+  private static final String PROCESS_ID = "process";
+
+  @Rule public final EngineRule engine = EngineRule.singlePartition();
+
+  @Test
+  public void shouldRejectActivateIfFlowScopeIsActivating() {
+    // given (synthetic situation - is not expected in regular processing)
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/processinstance/ProcessInstanceCommandRejectionTest.java:38]: I added some test cases which should not occur in regular processing. However, the cases are covered by the guard to protect the engine from bugs.

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java [hunk 2]
```diff
@@ -90,32 +54,18 @@ public boolean isValidStateTransition(final BpmnElementContext context) {
...
-
-      case ACTIVATE_ELEMENT:
-      case SEQUENCE_FLOW_TAKEN:
-        return hasActiveFlowScopeInstance(context);
-
       default:
         return Either.left(
             String.format(
-                "Expected event to have a process instance intent but was '%s'",
+                "Expected the check the preconditions of a command [activate,complete,terminate] but the intent was '%s'",
```
Comment (pihme) [inline engine/src/main/java/io/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java:68]:
```suggestion
                "Expected the check of the preconditions of a command with intent [activate,complete,terminate] but the intent was '%s'",
```

**Review by pihme — APPROVED** [review]
:tada:

**Comment by saig0** [comment]
@pihme thanks for the lightning-fast review :zap:

**Pull request merged**